py/src/nupic/optimization/swarming.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/python3
+# ------------------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+#
+# Copyright (C) 2018-2019, David McDougall
+#
+# This program is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Affero Public License version 3 as published by the Free
+# Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License along with
+# this program.  If not, see http://www.gnu.org/licenses.
+# ------------------------------------------------------------------------------
+""" Swarming parameter search """
+
+# TODO: Deal with global constants: particle_strength, global_strength, velocity_strength
+#       Maybe make them into CLI Arguments?
+
+particle_strength   =  .25
+global_strength     =  .50
+velocity_strength   =  .95
+assert(velocity_strength + particle_strength / 2 + global_strength / 2 >= 1)
+
+import argparse
+import sys
+import os
+import random
+import pprint
+
+from .nupic.optimization.parameter_set import ParameterSet
+
+class ParticleSwarmOptimizations:
```

Comment:
I do really like the recent changes! :clap: , extracting the particular search mechanism away, and adding the new PSO.

Reply:
> extracting the particular search mechanism away

This turned out to be more problematic than I expected.  Each search method needs:
1) Command line arguments
2) Queue experiments
3) Collect results, asynchronously from the queuing them

Currently it can only queue experiments.  I'm thinking about how to do all of this in a sane way that allows for new methods to plug-in.  I might make a new interface class to help with this.

Reply:
have you considered the 3rd party for param optimization, model selection? And "just" write enabling interface for our params & constructors?

Reply:
> https://www.khronos.org/nnef

I don't think this framework is useful, it's all about sharing trained deep learning networks.

> https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

Scikit learn appears very limited. 
- The only search method it has is GridSearch.  It does not do swarming.
- I also want the parameter optimization code to save all of the outputs of the program.  Which scikitlearn doesn't do.

Reply:
agreed on NNEF 

> Scikit learn appears very limited. 

it was only a first hit I got from memory, these "python parameter optimization frameworks" look interesting: 
https://hyperopt.github.io/hyperopt/ 
https://optuna.org/ 
I do like your solution, but in the longrun I hope we'll have an interface to a 3rd party project, as there's lots of details which are out of scope for us (same as Serialization, ..)

Reply:
> have an interface to a 3rd party project, as there's lots of details which are out of scope for us

I understand, I will do some more research into existing solutions.

Reply:
> 3rd party solutions

I found a great project, it's called [nni](https://github.com/Microsoft/nni).  It is a framework for optimizing artificial neural networks (the non-biological kind).  I read through it and attempted (but ultimately failed) to run the MNIST example with it.
- Its open source (MIT license) so we could easily incorporate it into our project.
- It's run by microsoft, and they appear to be actively developing it. They accept external PR's and they might even fix issues we have with it.
- It is well designed.  It's got some neat features too.

All that praise said, I don't think it's ready for mainstream use.  It's rough at the edges.  The project is about a year old.  In a year this could be the defacto standard tool, but today I'd rather keep using my tools which I'm familiar with.

Reply:
> I found a great project, it's called nni.  

I like NNI! although the whole AutoML seems overkill for us (we don't need network architecture "evolution", as we have HTM, we "just" need a way for param optimization. 

> I read through it and attempted (but ultimately failed) to run the MNIST example with it.

ok, let's go with what you have now. I would suggest not "wasting" time developing too many features, as long term we'd like to switch over to an external framework to do the work for us. 

The current will be a great improvement, so let's :+1:  

PS: what have failed running their mnist? Do you think it's not ready yet, or should I try in some later PR? 

Ideally, if you can take their (NNI or other suitable) design in mind, and develop the format for "experiment description" what you call it, similar/same, so we can later easily switch?

Reply:
> PS: what have failed running their mnist?

I tried running our MNIST in their framework.

> the format for "experiment description"

This is highly dependent on the specifics of the optimization framework.

Reply:
I will push the work i did for NNI in a branch so that you can play with it.

I think the problem i had is that NNI does not start by trying the default parameters.  It spent a considerable amount of time trying parameters which obviously wont work.  Swarming has an advantage over TPE (the search method nni defaults to) in that it starts near the default parameter values which are at least sane.  The NNI framework has a way to end trials/experiments early if they obviously arent working, but i didnt use it.